### PR TITLE
MSC4242: State DAGs (inbound joins)

### PR DESCRIPTION
**Pull request by kegsay**
Process inbound requests to join a state DAG room.

Split out from https://github.com/element-hq/synapse/pull/19425

Part of a series of 5x PRs to land the federation part of MSC4242 (https://github.com/element-hq/synapse/pull/19718, https://github.com/element-hq/synapse/pull/20127, https://github.com/element-hq/synapse/pull/20133, inbound-joins (this PR), inbound-pulls).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/federation_event.py [hunk 1]
```diff
@@ -555,6 +558,13 @@ async def process_remote_join(
         if room_version.identifier != room_version_id:
             raise SynapseError(400, "Room version mismatch")
 
+        if room_version.msc4242_state_dags and not is_state_dag_connected(
+            [ev for ev in state if supports_msc4242_state_dag(ev)]
+        ):
+            # We should have been given a connected state DAG. If there is a gap in it we
+            # cannot calculate the state, so the response is invalid and we refuse the join.
+            raise SynapseError(502, "State DAG is not connected")
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]:
```suggestion
            raise SynapseError(502, "Unable to join because the remote server passed back a state DAG that is not connected (invalid)")
```

File: synapse/handlers/federation_event.py [hunk 2]
```diff
@@ -566,9 +576,14 @@ async def process_remote_join(
...
-        await self._auth_and_persist_outliers(
-            room_id, itertools.chain(auth_events, state)
+        has_rejected_events = await self._auth_and_persist_outliers(
+            room_id,
+            itertools.chain(auth_events, state),
+            from_send_join=True,
         )
+        if room_version.msc4242_state_dags and has_rejected_events:
+            # The state DAG must not include rejected events
+            raise SynapseError(502, "State DAG included rejected events")
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]:
```suggestion
            raise SynapseError(502, "Unable to join because the remote server passed back a state  DAG that includes rejected events (invalid)")
```

File: synapse/handlers/federation_event.py [hunk 3]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                        if event_id in event_map
+                    }
+                    missing_auth_event_ids = set(calculated_auth_event_ids).difference(
+                        calculated_auth_events
+                    )
+                    if missing_auth_event_ids:
+                        calculated_auth_events.update(
+                            await self._store.get_events(
+                                missing_auth_event_ids,
+                                allow_rejected=True,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1990]: Feel like we should comment why we `allow_rejected=True` in both spots here

File: synapse/handlers/federation_event.py [hunk 4]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                                redact_behaviour=EventRedactBehaviour.as_is,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1991]: Comment why `redact_behaviour=EventRedactBehaviour.as_is`

File: synapse/handlers/federation_event.py [hunk 5]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                            )
+                        )
+                    batched_auth_events = {
+                        event_id: event_map[event_id]
+                        for event_id in itertools.chain(
+                            calculated_auth_event_ids, event.prev_state_events
+                        )
+                        if event_id in event_map
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2002]: Shouldn't it be an error if we don't have the event 🤔

File: synapse/handlers/federation_event.py [hunk 6]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
                 backfilled=True,
             )
 
+        return has_rejected_events
+
+    async def _auth_and_persist_state_dag_outliers(
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1896]: It's hard for me to be totally confident in what this is doing so I'm leaning on you that the intricacy is correct.

File: synapse/handlers/federation_event.py [hunk 7]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                try:
+                    validate_event_for_room_version(event)
+                    await check_state_independent_auth_rules(
+                        self._store, event, batched_auth_events
+                    )
+                    check_state_dependent_auth_rules(
+                        event, calculated_auth_events.values()
+                    )
+                except AuthError as e:
+                    logger.warning("Rejecting %r because %s", event, e)
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]:
```suggestion
                    logger.warning("Rejecting %r while persisting state DAG because %s", event, e)
```

I guess the previous message is based on what we're already doing in `_auth_and_persist_outliers` 🤷

File: synapse/handlers/federation_event.py [hunk 8]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                    ),
+                )
+                state_ids = await res.get_state(self._state_storage_controller)
+        else:
+            state_ids = await self._state_handler.compute_state_after_events(
+                event.room_id,
+                event.prev_state_events,
+                # We cannot filter the state as we need to persist the state group.
+                state_filter=None,
+                await_full_state=False,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2141]: Comment why `await_full_state=False`

For example, here is how we explain it elsewhere: https://github.com/element-hq/synapse/blob/f7cd91f6b0dc5e771ef01caf1680c4ab3a572fcc/synapse/handlers/sliding_sync/__init__.py#L454-L462

File: synapse/handlers/federation_event.py [hunk 9]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                        self._store, self._state_deletion_store
+                    ),
+                )
+                state_ids = await res.get_state(self._state_storage_controller)
+        else:
+            state_ids = await self._state_handler.compute_state_after_events(
+                event.room_id,
+                event.prev_state_events,
+                # We cannot filter the state as we need to persist the state group.
+                state_filter=None,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: This comment doesn't explain why. I had a similar question about this [before](https://github.com/element-hq/synapse/pull/19425#discussion_r2880582537)

I assume it's because we want all state in order to get an accurate state group

File: synapse/handlers/federation_event.py [hunk 10]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+                await_full_state=False,
+            )
+
+        # We should always have some resolved state after the prev_state_events, except
+        # for the create event which is the start of the state DAG.
+        is_create_event = (
+            len(event.prev_state_events) == 0 and event.type == EventTypes.Create
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: We also need to check for `state_key == ""` (canonical create event)

File: synapse/handlers/federation_event.py [hunk 11]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
+        )
+        if not is_create_event:
+            assert len(state_ids) > 0
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2154]: Add assertion message

Previous review comment: https://github.com/element-hq/synapse/pull/19425#discussion_r2880584310

File: synapse/handlers/federation_event.py [hunk 12]
```diff
@@ -1826,13 +1885,262 @@ async def prep(event: EventBase) -> None:
...
     async def _check_event_auth(
         self, origin: str | None, event: EventBase, context: EventContext
     ) -> None:
         """
         Checks whether an event should be rejected (for failing auth checks).
 
+        For MSC4242 State DAG rooms the auth events are calculated from the state before
+        the event rather than taken from the event, so checking the event against its auth
+        events (step 4) is the same as checking it against the state before it (step 5),
+        and only step 4 is performed.
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]:
```suggestion
        For MSC4242 State DAG rooms, the auth events are calculated from the state before
        the event rather than taken from the event, so checking the event against its auth
        events (step 4) is the same as checking it against the state before it (step 5),
        and only step 4 is performed.
```

Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: Feels like this comment is better placed below on the early return

File: synapse/handlers/federation_event.py [hunk 13]
```diff
@@ -2193,6 +2519,34 @@ async def _load_or_fetch_auth_events_for_event(
...
+
+        Returns:
+            The calculated auth events.
+        """
+        state_ids = await context.get_prev_state_ids()
+        calculated_auth_event_ids = self._event_auth_handler.compute_auth_events(
+            event, state_ids
+        )
+        event.internal_metadata.calculated_auth_event_ids = calculated_auth_event_ids
+        calculated_auth_events = await self._store.get_events(calculated_auth_event_ids)
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: Perhaps we be explicit about `allow_rejected=False` given the docstring note about that

File: synapse/handlers/federation_event.py (hunk 6 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:1896]: This isn't ideal. I'll try to refine/reword this so you can have more confidence.

File: synapse/handlers/federation_event.py (hunk 4 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:1991]: It's the same as the non state DAG path, so not really sure why it deserves special treatment.

Conceptually, both `as_is` and `redact` should work fine as the fields necessary for auth are protected from redaction. So long as they aren't `block`, as that would cause the event to be dropped and thus auth checks can't be performed.

File: synapse/handlers/federation_event.py (hunk 13 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

File: synapse/handlers/federation_event.py (hunk 12 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

File: synapse/handlers/federation_event.py (hunk 11 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2154]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

File: synapse/handlers/federation_event.py (hunk 10 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

File: synapse/handlers/federation_event.py (hunk 9 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

You're right in that there is no filter to apply, so "cannot filter" is just misleading. Reworded.

File: synapse/handlers/federation_event.py (hunk 3 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:1990]: https://github.com/element-hq/synapse/pull/20194/changes/86ac2c3dd3f82f220507abfa37ff59afd5bfabf3

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2002]: It is, in `check_state_independent_auth_rules`

File: synapse/handlers/federation_event.py (hunk 8 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2141]: https://github.com/element-hq/synapse/pull/20194/changes/1b7151ba87d6b4064752b61a1424cd774802975b

File: synapse/handlers/federation_event.py (hunk 6 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:1896]: https://github.com/element-hq/synapse/pull/20194/changes/1b7151ba87d6b4064752b61a1424cd774802975b

File: changelog.d/20194.misc [hunk 14]
```diff
@@ -0,0 +1 @@
+Add support for joining [MSC4242](https://github.com/matrix-org/matrix-spec-proposals/pull/4242) rooms over federation.
```
Comment (MadLittleMods) [inline changelog.d/20194.misc:1]: (dummy comment to start a review, resolve me)

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: If we're assuming the events should be in the `event_map`, we should assert that assumption instead of silently violating it.

In this case, that would be mean getting rid of `if event_id in event_map` and relying on `event_map[event_id]` blowing up

File: synapse/handlers/federation_event.py (hunk 9 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: > `# There is no filter to apply here`

This is right but it seems like the half the story. We want all of the events regardless.

File: synapse/handlers/federation_event.py [hunk 15]
```diff
@@ -1826,6 +1891,277 @@ async def prep(event: EventBase) -> None:
...
+                # the entire state DAG. To wait here would be circular, we'd deadlock.
+                await_full_state=False,
+            )
+
+        # We should always have some resolved state after the prev_state_events, except
+        # for the create event which is the start of the state DAG.
+        is_create_event = (
+            len(event.prev_state_events) == 0
+            and event.type == EventTypes.Create
+            and event.get_state_key() == ""
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: Alternative:

```suggestion
            and (event.type, event.get_state_key()) == (EventTypes.Create, "")
```

File: synapse/handlers/federation_event.py (hunk 11 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: I was thinking an actual message for when the assertion fails (not a comment)

File: synapse/handlers/federation_event.py [hunk 16]
```diff
@@ -2193,6 +2551,36 @@ async def _load_or_fetch_auth_events_for_event(
...
+    ) -> Collection[EventBase]:
+        """Calculate the auth events for an MSC4242 State DAG event and load them.
+
+        MSC4242 events do not list their auth events, so they are calculated from the
+        state before the event, in the same way `EventBuilder.build` does when creating
+        events locally. The calculated auth event IDs are remembered on the event so we
+        don't need to calculate them again when the event is persisted.
+
+        Rejected events never form part of the room state, so the calculated auth events
+        cannot be rejected and are loaded without allowing rejected events.
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: We should move this comment next to `allow_rejected=False`

File: synapse/handlers/federation_event.py [hunk 17]
```diff
@@ -1826,6 +1891,277 @@ async def prep(event: EventBase) -> None:
...
+        Raises:
+            SynapseError if the event has no `prev_state_events` and is not a create
+            event.
+        """
+        if len(event.prev_state_events) == 0 and event.type != EventTypes.Create:
+            raise SynapseError(502, f"event {event.event_id} has no prev_state_events")
+
+        if known_prev_state_maps:
+            if len(known_prev_state_maps) == 1:
+                state_ids = list(known_prev_state_maps.values())[0]
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2120]: What's the assumption being made here? (comment)

File: synapse/handlers/federation_event.py (hunk 8 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2143]: This comment was added:

```python
                # we need to have finished processing the entire state DAG before we can
                # compute the state for new events, but this function is called WHEN we are processing
                # the entire state DAG. To wait here would be circular, we'd deadlock.
                await_full_state=False,
```

But this makes me question how `compute_state_after_events` can even be accurate then?

File: synapse/handlers/federation_event.py (hunk 17 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2120]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: synapse/handlers/federation_event.py (hunk 16 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: synapse/handlers/federation_event.py (hunk 15 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: synapse/handlers/federation_event.py (hunk 11 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: synapse/handlers/federation_event.py (hunk 9 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: synapse/handlers/federation_event.py (hunk 8 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2143]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

Reworded to hopefully be clearer. I was too loose before when saying "entire" when it should have been "causally prior"

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/f466b2f52b0ffc8ea07de3c2a710f155d84f8628

File: changelog.d/20194.misc (hunk 14 above)
Comment (MadLittleMods) [reply changelog.d/20194.misc:1]: (another review)

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: `if event_id in event_map` was still left behind in one spot

File: synapse/handlers/federation_event.py [hunk 18]
```diff
@@ -1826,6 +1891,278 @@ async def prep(event: EventBase) -> None:
...
+               been persisted yet.
+        Returns:
+            The persisted EventContext and the calculated auth event IDs.
+
+        Raises:
+            SynapseError if the event has no `prev_state_events` and is not a create
+            event.
+        """
+        if len(event.prev_state_events) == 0 and event.type != EventTypes.Create:
+            raise SynapseError(502, f"event {event.event_id} has no prev_state_events")
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2115]: We should also be checking for canonical create event here `(EventTypes.Create, "")`. Probably just need to hoist `is_create_event`

Previous conversation: https://github.com/element-hq/synapse/pull/20194#discussion_r3983514993, https://github.com/element-hq/synapse/pull/20194#discussion_r3992275938

File: synapse/handlers/federation_event.py (hunk 17 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2120]: This is the comment that was added:

```
# There's only 1 state map so we don't need to do state resolution at all.
```

Feels like a detail that could be part of `resolve_state_groups(...)` itself if this is possible.

And it looks like we're doing this more because the docstring of `resolve_state_groups(...)` tells us to:

https://github.com/element-hq/synapse/blob/8c8138d57247258f5e1ae90e52d01b90c7d140a0/synapse/state/__init__.py#L660-L661

File: synapse/handlers/federation_event.py (hunk 8 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2143]: Overall, I think it's unclear how/why `state_ids` is accurate if we're not taking into account the events in the State DAG prior to `event`.

Comment (kegsay) [reply synapse/handlers/federation_event.py:2146]: I don't understand how you can be unclear about this? `state_ids = await self._state_handler.compute_state_after_events(event.room_id, event.prev_state_events, ...)` literally is taking into account the state dag prior to `event`... that is what `event.prev_state_events` is doing.

File: synapse/handlers/federation_event.py (hunk 17 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2123]: I have no desire to change the behaviour of `resolve_state_groups` in this PR as it is unrelated to the purpose of this PR.

File: synapse/handlers/federation_event.py (hunk 18 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/51b9c9d7339f57517d6e4cfd9a7f5c85cade1fc0

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/51b9c9d7339f57517d6e4cfd9a7f5c85cade1fc0

File: changelog.d/20194.misc (hunk 14 above)
Comment (MadLittleMods) [reply changelog.d/20194.misc:1]: (another review)

File: synapse/handlers/federation_event.py (hunk 8 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2146]: I think what's confusing is the comment basically says that we haven't finished processing the entire state DAG but then also says we need to have done so; with no mention of whether we have done so or whether we've checked.

With a keen eye and assumptions, your latest correction is better but I'm only really picking this apart now with hindsight.

And we could re-frame it a bit:

```suggestion
                # By this point, we expect to have already finished processing the
                # events *causally prior* to `event` in the state DAG because this
                # function is called unilaterally WHEN we are processing events in the
                # state DAG all the way down the chain and the caller ensures we process
                # events in topological order.
                #
                # If this assumption is ever violated, `compute_state_after_events` will
                # yell loudly about missing state groups.
                #
                # In the context of partially stated rooms (faster remote room joins)
                # this would be part of background step verifying the entire state DAG,
                # so we know that we don't have the full state processed yet. To wait
                # here would be circular, we'd deadlock.
                await_full_state=False,
```

File: synapse/handlers/federation_event.py (hunk 17 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2123]: :+1: but we should update the comment to point out the reason more clearly (because `resolve_state_groups(...)` says so)

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: An LLM is pointing out that it's possible to trigger the `KeyError` for a remote invite which goes through `_auth_and_persist_outliers(...)` and [filters out events from the `event_map` that we already persisted](https://github.com/element-hq/synapse/blob/cf02d0d1d70fa26b415d30c792cbf1a3f44e90be/synapse/handlers/federation_event.py#L1716-L1724) before being passed to `_auth_and_persist_state_dag_outliers(...)`. Looks plausible in which case the comments need adjusting here. `event_map is the complete state DAG` isn't true.

Do we have a test for this?

File: synapse/handlers/federation_event.py
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1773]: LLM spotted this:

`have_seen_events(...)` doesn't take into account rejected events which means the `has_rejected_events` bool result that we return isn't accurate in the case where we've already persisted the rejected events.

File: synapse/handlers/federation_event.py [hunk 19]
```diff
@@ -1826,6 +1891,278 @@ async def prep(event: EventBase) -> None:
...
+                    }
+                    # If we have calculated state before this event, remember it so we don't need
+                    # to keep asking the database. The state we are remembering is StateMap[event_id]
+                    # along with the state group integer, which is exactly the data needed by
+                    # _calculate_state_dag_context
+                    if context.state_group is not None:
+                        event_id_to_state_group[event.event_id] = context.state_group
+                        state_group_to_state_map[
+                            context.state_group
+                        ] = await context.get_current_state_ids()
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:2010]: LLM is mentioning that we store the `state_group` here before checking the auth rules. Which means if an event turns out to be rejected, it's still ends up as part of all descendants.

File: synapse/handlers/federation_event.py [hunk 20]
```diff
@@ -1826,6 +1891,278 @@ async def prep(event: EventBase) -> None:
...
+               `prev_state_events` come before it
+            from_send_join: True if `event_map` is the complete state DAG from a
+               /send_join response
+        Returns:
+            True if any of the events were rejected.
+        """
+        has_rejected_events = False
+        events_and_contexts_to_persist: list[EventPersistencePair] = []
+        event_id_to_state_group: dict[str, int] = {}
+        state_group_to_state_map: dict[int, StateMap[str]] = {}
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1929]: Since the state DAG can have thousands of events and we store a whole `StateMap` for each event, LLM is suggesting we could reduce the memory foot-print here by evicting an entry once all of its successors are processed. Or capping the size and falling back to the DB to bound it.

Seems like a future optimization ⏩

File: synapse/handlers/federation_event.py [hunk 21]
```diff
@@ -566,9 +579,17 @@ async def process_remote_join(
...
+        has_rejected_events = await self._auth_and_persist_outliers(
+            room_id,
+            itertools.chain(auth_events, state),
+            from_send_join=True,
         )
+        if room_version.msc4242_state_dags and has_rejected_events:
+            # The state DAG must not include rejected events
+            raise SynapseError(
+                502,
+                "Unable to join because the remote server passed back a state  DAG that includes rejected events (invalid)",
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:591]:
```suggestion
                "Unable to join because the remote server passed back a state DAG that includes rejected events (invalid)",
```

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: This is addressed in the invite PR https://github.com/element-hq/synapse/pull/20226/changes#diff-a91beeaf44154c199a11f5539f2358704a1df966ff74ad8f2681b64430843f8cR2071

You currently can't even get that far in the invite-join dance since the invite itself fails over federation.

File: synapse/handlers/federation_event.py (hunk 20 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:1936]: Yeah falling to DB to bound it would be my approach, but not for now.

File: synapse/handlers/federation_event.py (hunk 19 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2017]: I think this is wrong.

We do plop the event in the `event_id_to_state_group` map prior to checking auth rules, but so what? "it's still ends up as part of all descendants." is nonsensical. Descendants need to explicitly reference that event in `prev_state_events`, in which case it would be cascade rejected anyway according to state DAG auth rules, so I see no issue here.

File: synapse/handlers/federation_event.py (hunk 17 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2131]: https://github.com/element-hq/synapse/pull/20194/changes/ab6b28323a562df736929cad169e5b23f78baaa5

File: synapse/handlers/federation_event.py
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/ab6b28323a562df736929cad169e5b23f78baaa5

File: synapse/handlers/federation_event.py [hunk 22]
```diff
@@ -1826,6 +1898,292 @@ async def prep(event: EventBase) -> None:
...
+                # events in topological order.
+                #
+                # If this assumption is ever violated, `compute_state_after_events` will
+                # yell loudly about missing state groups.
+                #
+                # In the context of partially stated rooms (faster remote room joins)
+                # this would be part of background step verifying the entire state DAG,
+                # so we know that we don't have the full state processed yet. To wait
+                # here would be circular, we'd deadlock.
+                await_full_state=False,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: Old comment is still here

```suggestion
                # By this point, we expect to have already finished processing the
                # events *causally prior* to `event` in the state DAG because this
                # function is called unilaterally WHEN we are processing events in the
                # state DAG all the way down the chain and the caller ensures we process
                # events in topological order.
                #
                # If this assumption is ever violated, `compute_state_after_events` will
                # yell loudly about missing state groups.
                #
                # In the context of partially stated rooms (faster remote room joins)
                # this would be part of background step verifying the entire state DAG,
                # so we know that we don't have the full state processed yet. To wait
                # here would be circular, we'd deadlock.
                await_full_state=False,
```

Context: https://github.com/element-hq/synapse/pull/20194#discussion_r3983499574

File: synapse/storage/databases/main/rejections.py [hunk 23]
```diff
@@ -35,3 +35,15 @@ async def get_rejection_reason(self, event_id: str) -> str | None:
             allow_none=True,
             desc="get_rejection_reason",
         )
+
+    async def get_rejected_events(self, event_ids: set[str]) -> set[str]:
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/rejections.py]: Perhaps `has_rejected_event_ids`

We already have `have_seen_events`, `have_censored_event`

File: synapse/handlers/federation_event.py [hunk 24]
```diff
@@ -1719,35 +1768,53 @@ async def _auth_and_persist_outliers(
...
         # This is just an optimisation, so it doesn't need to be watertight - the event
         # persister does another round of deduplication.
+        has_rejected_events = False
         seen_remotes = await self._store.have_seen_events(room_id, event_map.keys())
-        for s in seen_remotes:
-            event_map.pop(s, None)
+        if seen_remotes:
+            for s in seen_remotes:
+                event_map.pop(s, None)
+            rejected_event_ids = self._store.get_rejected_events(seen_remotes)
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py]: Missing `await`

```suggestion
            rejected_event_ids = await self._store.get_rejected_events(seen_remotes)
```

### Test

Do we at-least have a happy-path test for previously persisted events?

A test for this edge case would be nice.


### Possible lints

Too bad [`unused-coroutine`](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-coroutine-return-value-is-used-unused-coroutine) doesn't catch this

The [`truthy-bool`](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-expression-is-not-implicitly-true-in-boolean-context-truthy-bool) mypy lint might have caught this

File: synapse/handlers/federation_event.py (hunk 19 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:2017]: I guess the concern here is just to double check that we're not successfully authing events with rejected state :thinking:

LLM does come up with some tests to reproduce the problem but I think you could see the same if you asked the LLM yourself so I won't share the unvettted test. Since I haven't fully onboarded to the test, unclear if its cheating to get it to fail. The `SJ02` [Complement test](https://github.com/matrix-org/complement/pull/841) is in the right vein of stress testing this but only checks for the `502` failure from the join itself, not what it leaves behind and authing further events.

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: Looking at the code again, I think this is also applicable with a remote join as well. Just any time we've already persisted some of the events and pop things out of the `event_map`

File: synapse/storage/databases/main/rejections.py [hunk 25]
```diff
@@ -35,3 +35,15 @@ async def get_rejection_reason(self, event_id: str) -> str | None:
...
+        """Filter the provided event IDs to only return rejected events."""
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/rejections.py]: We should note that this doesn't check whether we have this event persisted

As a missing event, would appear as a non-rejected event

File: synapse/handlers/federation_event.py (hunk 24 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: I'm very surprised tests didn't catch this, so will add an integration test. I would need to join the room -> leave -> rejoin for `have_seen_events` to return something.

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: Hmm, I agree I think you're right. I _think_ this will be tested with the re-join integration test, given we would pop the state events we'd seen the first time round, so let's see. 

I do want to keep the invariant of `event_map == complete state DAG if from_send_join` as it makes the processing logic easier to understand, so the correct fix is to ensure we don't pop for state DAG rooms.

File: synapse/handlers/federation_event.py (hunk 19 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py:2017]: I'll add a test to check what happens if the state DAG has a rejected event and then a single otherwise valid event coming after it (it should cascade reject).

File: synapse/storage/databases/main/rejections.py (hunk 23 above)
Comment (kegsay) [reply synapse/storage/databases/main/rejections.py]: https://github.com/element-hq/synapse/pull/20194/changes/652fdbc29294a54929dc1c7d4031b02fd79cdf83

File: synapse/storage/databases/main/rejections.py (hunk 25 above)
Comment (kegsay) [reply synapse/storage/databases/main/rejections.py]: https://github.com/element-hq/synapse/pull/20194/changes/652fdbc29294a54929dc1c7d4031b02fd79cdf83

File: synapse/handlers/federation_event.py (hunk 24 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: https://github.com/element-hq/synapse/pull/20194/changes/652fdbc29294a54929dc1c7d4031b02fd79cdf83 adds the `await`, will add tests now and verify that removing `await` makes it fail.

File: synapse/handlers/federation_event.py (hunk 5 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: Tests: https://github.com/matrix-org/complement/pull/926
Fix: https://github.com/element-hq/synapse/pull/20194/changes/ac6af6bc3cf3d119bd0a025e0da3e6ee281e3aa4

File: synapse/handlers/federation_event.py (hunk 19 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: Tests: https://github.com/matrix-org/complement/pull/926
Fix: https://github.com/element-hq/synapse/pull/20194/changes/ac6af6bc3cf3d119bd0a025e0da3e6ee281e3aa4

Turns out you were right, we would 502 due to having rejected events but incorrectly accept the rejected event.

File: synapse/handlers/federation_event.py (hunk 24 above)
Comment (kegsay) [reply synapse/handlers/federation_event.py]: Tests: https://github.com/matrix-org/complement/pull/926

Verified dropping the await makes the complement tests  fail... but had to do https://github.com/element-hq/synapse/pull/20194/changes/a21d1da3de058ae087f7b70aaf84f2cd649558ed to ensure that.
